### PR TITLE
355 execute command in terminal returns wrong output on windows

### PR DESCRIPTION
**Pull request by JanPrusinowski**

### CODE REVIEW

File: self-tests/terminal.robot [hunk 1]
```diff
@@ -37,3 +37,28 @@ Execute UEFI Shell Command
...
+Execute Command In Terminal over SSH
+    [Documentation]    Tesest Execute Command In Terminal kwd over SSH
+    Power On
+    Login To Windows
+    Set Test Variable    ${COMMAND}    Get-PnpDevice -PresentOnly | Where-Object { $_.InstanceId -match '^USB' }
+    ${out1}=    Execute Command In Terminal    ${COMMAND}
+    ${out2}=    Execute Command In Terminal    ${COMMAND}
+    Should Be Equal As Strings    ${out1}    ${out2}
+    Should Not Contain    ${out1}    ${COMMAND}
+    Should Not Contain    ${out1}    ${COMMAND}
```
Comment (m-iwanicki) [inline self-tests/terminal.robot:50]:
```suggestion
    Should Not Contain    ${out2}    ${COMMAND}
```

File: self-tests/terminal.robot [hunk 2]
```diff
@@ -37,3 +37,28 @@ Execute UEFI Shell Command
...
+
+    Set Test Variable    ${COMMAND}    ls
+    ${out1}=    Execute Command In Terminal    ${COMMAND}
+    ${out2}=    Execute Command In Terminal    ${COMMAND}
+    Should Be Equal As Strings    ${out1}    ${out2}
+    Should Not Contain    ${out1}    ${COMMAND}
+    Should Not Contain    ${out1}    ${COMMAND}
```
Comment (m-iwanicki) [inline self-tests/terminal.robot:57]:
```suggestion
    Should Not Contain    ${out2}    ${COMMAND}
```

File: self-tests/terminal.robot [hunk 3]
```diff
@@ -37,3 +37,28 @@ Execute UEFI Shell Command
...
+
+    Set Test Variable    ${COMMAND}    driverquery
+    ${out1}=    Execute Command In Terminal    ${COMMAND}
+    ${out2}=    Execute Command In Terminal    ${COMMAND}
+    Should Be Equal As Strings    ${out1}    ${out2}
+    Should Not Contain    ${out1}    ${COMMAND}
+    Should Not Contain    ${out1}    ${COMMAND}
```
Comment (m-iwanicki) [inline self-tests/terminal.robot:64]:
```suggestion
    Should Not Contain    ${out2}    ${COMMAND}
```

File: lib/terminal.robot [hunk 4]
```diff
@@ -170,5 +172,6 @@ Execute UEFI Shell Command
     Write Bare Into Terminal    ${command}
     Sleep    ${timeout}ms
     Press Enter
+    Sleep    1s
```
Comment (m-iwanicki) [inline lib/terminal.robot:175]: Is sleep needed?

**Review by m-iwanicki — CHANGES_REQUESTED** [review]

File: lib/terminal.robot [hunk 5]
```diff
@@ -172,3 +175,37 @@ Execute UEFI Shell Command
...
+    
+    FOR    ${line}    IN    @{lines}    
+        Run Keyword If    ${add}    Append To List    ${filtered_lines}    ${line}
+        Run Keyword If    ${add}    Log To Console    Added line: ${line}
+        
+        IF    ${add} == ${FALSE}
+            Log To Console    \nSTRING CONTAINS:
+            Log To Console    Line: ${line}
+            Log To Console    Command: ${command}
+            ${contains_command}=    Run Keyword And Return Status    String Should Contain    ${line}    ${command}
```
[the remark this answers is not in the document]
Comment (m-iwanicki) [reply lib/terminal.robot]: Outdated, forgot to send earlier.

File: lib/terminal.robot (hunk 4 above)
Comment (JanPrusinowski) [reply lib/terminal.robot:175]: While testing  a couple of times the command have not have a chance to start producing any output before the next keyword was being executed (the one collecting output). The outcome was an empty string as output.

File: lib/terminal.robot (hunk 5 above)
[the remark this answers is not in the document]
Comment (JanPrusinowski) [reply lib/terminal.robot]: This was just checking of a concept and trying to find a reason of the strange behaviour. Was fixed in a different way.

File: self-tests/terminal.robot [hunk 6]
```diff
@@ -37,3 +37,28 @@ Execute UEFI Shell Command
     Should Contain    ${out}    Device Name
     ${out}=    Execute UEFI Shell Command    bcfg boot dump
     Should Contain    ${out}    Optional- N
+
+Execute Command In Terminal over SSH
+    [Documentation]    Tesest Execute Command In Terminal kwd over SSH
```
Comment (m-iwanicki) [inline self-tests/terminal.robot:42]:
```suggestion
    [Documentation]    Test Execute Command In Terminal kwd over SSH
```

**Review by m-iwanicki — CHANGES_REQUESTED** [review]
https://github.com/Dasharo/open-source-firmware-validation/actions/runs/10106691593/job/27949252740

```
Execute Command In Terminal over SSH :: Tesest Execute Command In ... | FAIL |
'1 > 1' should be true.
```

Not sure if it's possible to somehow get `log.html` out of runner, if not then maybe it would be good enhancement to more easily debug those kind of errors.
Maybe there is no windows installed on runner QEMU?

**Comment by macpijan** [comment]
> Not sure if it's possible to somehow get log.html out of runner, if not then maybe it would be good enhancement to more easily debug those kind of errors.

that is being done here: https://github.com/Dasharo/open-source-firmware-validation/pull/410

> Maybe there is no windows installed on runner QEMU?

It is not. @JanPrusinowski This test should not be executed in CI right now, as there is no OS in what we have right now.

**Comment by JanPrusinowski** [comment]
@macpijan I have excluded this test from run if run on qemu.

**Review by macpijan — CHANGES_REQUESTED** [review]

File: self-tests/terminal.robot (hunk 6 above)
Comment (m-iwanicki) [reply self-tests/terminal.robot:42]: Your force pushed version with error again here: https://github.com/Dasharo/open-source-firmware-validation/compare/65073119775eb665f7a476772415ba69ad6e8e95..2febf8c1f809312d17952acee6ea144f66c410e1

File: self-tests/terminal.robot (hunk 3 above)
Comment (m-iwanicki) [reply self-tests/terminal.robot:65]: Your force pushed version with error again here: https://github.com/Dasharo/open-source-firmware-validation/compare/65073119775eb665f7a476772415ba69ad6e8e95..2febf8c1f809312d17952acee6ea144f66c410e1

File: self-tests/terminal.robot (hunk 2 above)
Comment (m-iwanicki) [reply self-tests/terminal.robot:58]: Your force pushed version with error again here: https://github.com/Dasharo/open-source-firmware-validation/compare/65073119775eb665f7a476772415ba69ad6e8e95..2febf8c1f809312d17952acee6ea144f66c410e1

File: self-tests/terminal.robot (hunk 1 above)
Comment (m-iwanicki) [reply self-tests/terminal.robot:51]: @JanPrusinowski  Your force pushed version with error again here: https://github.com/Dasharo/open-source-firmware-validation/compare/65073119775eb665f7a476772415ba69ad6e8e95..2febf8c1f809312d17952acee6ea144f66c410e1

File: self-tests/terminal.robot [hunk 7]
```diff
@@ -37,3 +37,29 @@ Execute UEFI Shell Command
...
```
[the remark this answers is not in the document]
Comment (JanPrusinowski) [reply self-tests/terminal.robot]: added this link to the documentation: https://github.com/Dasharo/open-source-firmware-validation/issues/355

File: self-tests/terminal.robot [hunk 8]
```diff
@@ -37,3 +37,29 @@ Execute UEFI Shell Command
...
+    Skip If    '${CONFIG}' == 'qemu'
```
[the remark this answers is not in the document]
Comment (JanPrusinowski) [reply self-tests/terminal.robot]: changed in: de67fa9154e2ea8e2c8a4eb894b7efd07fb1d1e4

File: self-tests/terminal.robot [hunk 9]
```diff
@@ -37,3 +37,29 @@ Execute UEFI Shell Command
     Should Contain    ${out}    Device Name
     ${out}=    Execute UEFI Shell Command    bcfg boot dump
     Should Contain    ${out}    Optional- N
+
+Execute Command In Terminal over SSH
```
[the remark this answers is not in the document]
Comment (JanPrusinowski) [reply self-tests/terminal.robot]: @macpijan added (Windows) at: de67fa9154e2ea8e2c8a4eb894b7efd07fb1d1e4

**Comment by macpijan** [comment]
Replaced with: https://github.com/Dasharo/open-source-firmware-validation/pull/429/files

**Pull request closed without merging**